File: 2024/2024-regexp/main.go
package main

import (
	"fmt"

	"github.com/dlclark/regexp2"
)

func main() {
	fmt.Printf("Go Code\n")
	PrintMatchOrNot(`a[bc]`, `ab`)
	PrintMatchOrNot(`a[bc]`, `ad`)
	PrintMatchOrNot(`a(?!b)`, `a`)
	PrintMatchOrNot(`a(?!b)`, `ab`)
	PrintMatchOrNot(`a(?!b)`, `ac`)
	PrintMatchOrNot(`a+(?=bc)`, `aaaadabcaadbc`)
	PrintMatchOrNot(`a+(?!bc)`, `abcaadbc`)
	PrintMatchOrNot(`\s+(?!\S)`, `World   `)
	PrintMatchOrNot(`\s+(?:[^\S])`, "World   ")
	PrintMatchOrNot(`\s+(?!\S)`, `   World`)
	PrintMatchOrNot(`\s+(?:[^\S])`, "   World")
	PrintMatchOrNot(`\s+(?!\S)`, `Hello   World  `)
	PrintMatchOrNot(` ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+`, `Hello   World  `)
	PrintMatchOrNot(`\s+(?:[^\S])`, "Hello   World  ")
	PrintMatchOrNot(`\s+`, "Hello   World  ")
}

func PrintMatchOrNot(pat, str string) {
	re := regexp2.MustCompile(pat, 0)
	m, _ := re.FindStringMatch(str)
	if m == nil {
		fmt.Printf("%5v With pat `%v` and str `%v`\n", "false", pat, str)
		return
	}

	idx := m.Groups()[0].Capture.Index
	l := m.Groups()[0].Capture.Length
	fmt.Printf("%5v With pat `%v` and str `%v`. Group: `%v` idx: %d len: %v\n", "true", pat, str, m.String(), idx, l)
}

File: 2024/2024-regexp/rs/src/main.rs
use fancy_regex::Regex;

fn main() {
    println!("Rust Code\n");
    print_match_or_not(r"a[bc]", "ab");
    print_match_or_not(r"a[bc]", "ad");
    print_match_or_not(r"a(?!b)", "a");
    print_match_or_not(r"a(?!b)", "ab");
    print_match_or_not(r"a(?!b)", "ac");
    print_match_or_not(r"a+(?=bc)", "aaaadabcaadbc");
    print_match_or_not(r"a+(?!bc)", "abcaadbc");
    print_match_or_not(r"\s+(?!\S)", "World   ");
    print_match_or_not(r"\s+(?:[^\S])", "World   ");
    print_match_or_not(r"\s+(?!\S)", "   World");
    print_match_or_not(r"\s+(?:[^\S])", "   World");
    print_match_or_not(r"\s+(?!\S)", "Hello   World  ");
    print_match_or_not(
        r" ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+",
        "Hello   World  ",
    );
    print_match_or_not(r"\s+(?:[^\S])", "Hello   World  ");
    print_match_or_not(r"\s+", "Hello   World  ");
}

fn print_match_or_not(pat: &'static str, s: &'static str) {
    let re = Regex::new(pat).unwrap();
    let result = re.find(s);

    let match_option = result.unwrap();

    if !match_option.is_some() {
        println!("false With pat `{}` and str `{}`", pat, s);
        return;
    }
    let m = match_option.unwrap();

    println!(
        " true With pat `{}` and str `{}`. Group: `{}` idx: {} len: {}",
        pat,
        s,
        m.as_str(),
        m.start(),
        m.end() - m.start()
    );
}
